File: kata/c#/Nir-Kata/Primitive.Obsession/NIR.cs
namespace Nir_Kata.Primitive.Obsession
{
    public static class NIR
    {
        private const int ValidLength = 15;

        public static bool Validate(string input) => ValidateLength(input)
                                                     && ValidateSex(input[0])
                                                     && ValidateYear(input[1..3])
                                                     && ValidateMonth(input[3..5])
                                                     && ValidateDepartment(input[5..7])
                                                     && ValidateCity(input[7..10])
                                                     && ValidateSerialNumber(input[10..13]);

        private static bool ValidateLength(string input) => input.Length == ValidLength;

        private static bool ValidateSex(char sex) => sex is '1' or '2';
        private static bool ValidateYear(string year) => year.IsANumber();

        private static bool ValidateMonth(string month) =>
            ValidateNumber(month, m => m is > 0 and <= 12);

        private static bool ValidateDepartment(string department) =>
            ValidateNumber(department, d => d is > 0 and <= 95 or 99);

        private static bool ValidateCity(string city) => city.IsANumber();

        private static bool ValidateSerialNumber(string serialNumber) => serialNumber.IsANumber();

        private static bool ValidateNumber(string potentialNumber, Predicate<int> predicate) =>
            potentialNumber
                .ToInt()
                .Match(number => predicate(number), false);
    }
}

File: kata/java/src/main/java/primitive/obsession/NIR.java
package primitive.obsession;

import lombok.experimental.ExtensionMethod;
import lombok.experimental.UtilityClass;

import java.util.function.Function;

@UtilityClass
@ExtensionMethod(StringExtensions.class)
public class NIR {

    private static final int VALID_LENGTH = 15;
    private static final char MALE = '1';
    private static final char FEMALE = '2';
    public static final int CONTROL_KEY_COMPLEMENT = 97;

    public static Boolean validate(String potentialNIR) {
        return validateLength(potentialNIR)
                && validateSex(potentialNIR.charAt(0))
                && validateYear(potentialNIR.substring(1, 3))
                && validateMonth(potentialNIR.substring(3, 5))
                && validateDepartment(potentialNIR.substring(5, 7))
                && validateCity(potentialNIR.substring(7, 10))
                && validateSerialNumber(potentialNIR.substring(10, 13))
                && validateKey(potentialNIR);
    }

    private static boolean validateKey(String potentialNir) {
        int sum = potentialNir
            .substring(0, 13)
            .chars()
            .map(Character::getNumericValue)
            .sum();
        int controlKey = CONTROL_KEY_COMPLEMENT - (sum % CONTROL_KEY_COMPLEMENT);
        return controlKey == Integer.parseInt(potentialNir.substring(13));
    }

    private static boolean validateLength(String potentialNIR) {
        return potentialNIR.length() == VALID_LENGTH;
    }

    private static boolean validateSex(char sex) {
        return sex == MALE || sex == FEMALE;
    }

    private static boolean validateYear(String year) {
        return year.isANumber();
    }

    private static boolean validateMonth(String month) {
        return validateNumber(month, x -> x > 0 && x <= 12);
    }

    private static boolean validateDepartment(String department) {
        return validateNumber(department, x -> x > 0 && (x <= 95 || x == 99));
    }

    private static boolean validateCity(String city) {
        return city.isANumber();
    }
    private static boolean validateSerialNumber(String serialNumber) {
        return serialNumber.isANumber();
    }
    private static boolean validateNumber(String potentialNumber, Function<Integer, Boolean> isValid) {
        return potentialNumber
                .toInt()
                .map(isValid)
                .getOrElse(false);
    }
}
